Adds History Feature with Updated Docs

Mike-Bros opened the pull request.

Commit:
Command History + New Demo

- Added functionality to page through command history using `shift` + `UP` or `DOWN`.
- Created new demo scene that tries to cover all of the uses currently available.

Commit:
- Applied consistent file formatting for gdscript.
- Added plugin description and version.
- Updated gitignore with Rider extensions and directories.
- Updated README

Commit:
Added screenshots to the README

KeilainMan commented on README.md:
```diff
@@ -5,19 +5,25 @@
 ## Installation
 
-Install via Godot AssetLib and enable plugin in project settings.
+[Install via Godot AssetLib](https://godotengine.org/asset-library/asset/2946) and enable plugin in project settings.
```
I really like that part :D

KeilainMan commented on addons/console_and_textchat/console.gd:
```diff
@@ -2,268 +2,389 @@
-		print_message(new_text)
+    if new_text == "":
+        return
+    _save_command_to_history(new_text)
```
I think it should be `_save_text_to_history`, as not everything is a command

Mike-Bros replied:
I agree, along with other instances of command/cli should really be updated to be clearer.

KeilainMan commented on addons/console_and_textchat/console.gd:
```diff
@@ -2,268 +2,389 @@
+        print_message(new_text)
 
-	text_input_line.text = ""
+    _clear_cli_text()
```
I guess for the sake of conformity, ´cli´ should be changed to `text_input_line` in **all** instances

KeilainMan commented on addons/console_and_textchat/demo/demo2.gd:
```diff
@@ -0,0 +1,63 @@
+    console_3.print_message("The list_args command was called with these arguments: " + console_3.italic(console_3.col(Color.CADET_BLUE, " ".join(PackedStringArray(args)))))
+
+
+func _on_console_3_toggle_pressed():
```
They look a bit unnecessary :D

KeilainMan commented on addons/console_and_textchat/console.gd:
```diff
@@ -2,268 +2,389 @@
-func print_message(text: String) -> void:
-	display_label.text += "\n"
-	display_label.text += text
+func print_message(text: String, codeblock_replacement = true) -> void:
```
I must admit, that I do not understand the purpose of changes in that function at all.

Mike-Bros replied:
I think this was a leftover from wanting to support some markdown-styled code blocks or something. I'm going to remove the changes to this function since it's half-baked and not used anywhere.

KeilainMan reviewed (changes requested):
**Hi!** Thank you for taking an interest in this small little addon!

Adding a chat history is an interesting feature, thank you for submitting this code. However, during the review, a few questions popped up in my head.

1. What is the purpose of the changed `print_message` function?
2. What were your thoughts behind the implemented themes logic? Maybe you can explain this a little bit.
3. Otherwise I made some comments to things that concerned consistency. Please change `cli` to `text_input_line` throughout the code. And change the other function mentioned in the comment. 
**Edit**
4. How about changing the inputkey from shift + up/down to only up/down? I think it is more intuitive.  

It is interesting to see, that your editor changed the indentation of all the code :D
**Edit** I just noticed, that Godot immediately reverts the indentation, when I open the script. 

Thank you for your efforts. If you have additional feedback, I am happy to hear about it!

Mike-Bros commented:
Thanks for your comments! I don't know when, but I expect I will get around to making another PR with the requested changes. By the way, it's an awesome little component. Thank you!

Summary response:

1. Not needed changes, will revert
2. The theme management is still a bit of a problem. The basic idea was that I wanted the text color to change when paging through the history. The theme is supplied to pass font coloring to the `LineEdit` for the command input. I need to rethink this, though I'm not happy with the current solution. That's why I hadn't written the updated docs for this part. I was actually hoping you might have some insight/ideas here.
3. 100% Agree and will do!
4. The problem here has to do with the [existing build-ins for the `LineEdit` node](https://docs.godotengine.org/en/stable/classes/class_lineedit.html). The up/down arrows move the caret to the beginning/end of the line without modifiers. I struggled with a compromise here, and I settled on the `shift` modifier so as not to break compatibility with the node. Alternatively, I also thought that maybe up/down would be more intuitive for history if I could swap the `LineEdit` caret reposition to use the shift modifier instead. Still, I'm new to Godot, so I am not sure how difficult it is to change that level of functionality on a node to accomplish this. I am interested in hearing your input on this if you have any ideas/preferences. I certainly prefer what I'm used to without the shift modifier, but can live with a modifier 🤷🏻.

KeilainMan commented:
Hey!

2. Mh... Indeed, the theme system is very complex for its purpose. As people can assign a theme to the node anyway, maybe it would be easier to `@export var font_color_history: Color`. When the user uses the history function the console enters `mode_history` which sets the color as font color. And this is reverted when the users leave the history mode. 
4. I see the problem. I read that ´_gui_input´ can overwrite the shortcuts. Also, there is a ´shortcut_enabled´ property.
    1. I never get the ´_gui_input` function to run. I don't know haha.. I will further check this out.
    2. The shortcut property did nothing in my tests, they still worked. Maybe they meant other shortcuts?

If you want I can offer you my help and see if I get a nice solution for 2. Regarding 4 I would say we leave it at shift + up/down and I will further investigate to find a solution for that. Ofc, if you find a way I welcome you to tell me :)

KeilainMan commented:
Hey how is it going? I would really like to implement the history feature :) Do you need some help?

Mike-Bros commented:
Hey sorry for the radio silence. I'm actually closing on a house this week
so I've been pretty caught up in that process and haven't done much Godot
work. I probably won't get back to this for at least a couple of weeks; if
you want to try implementing your own version of this in the meantime, I'd
be happy to help once I get some free time again!


On Mon, Sep 23, 2024 at 4:38 AM KeilainMan ***@***.***> wrote:

> Hey how is it going? I would really like to implement the history feature
> :) Do you need some help?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/KeilainMan/ConsoleAddon/pull/1#issuecomment-2367695024>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACHFB2ODJQIWKO37RKAM3D3ZX7OPNAVCNFSM6AAAAABI75CIZGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNRXGY4TKMBSGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

KeilainMan commented:
Sounds great , good luck with that! I think I will go ahead and make the necessary changes for your request to implement your solution with the things we talked about (at least for the history) but my time is also sparse, so lets see when it will happen. :)

KeilainMan committed:
Changes for history and themes

Code was changed and added as discussed. Themes were partly removed. User can picka highlighting color that will be applied due to the settings used

KeilainMan commented:
Thanks you for your contribution! I managed to go through the code and apply the changes we talked about. The history was added as you provided and I changed the Theme-Mechanic. You can apply a theme to the console and you can choose a highlighting color for the history. I found an interesting solution to not create new themes, but just override the coloring. Feel free to take a look into it if you are interested!

Thank you for your help! :)

Edit:
I was not able to find any solution to the Shift + Arrow Problem yet :( Maybe I ask reddit tomorrow :D
